### PR TITLE
Add support for `getUpgradeHandler` in custom server

### LINKED ISSUE #49334: [NEXT-1119] Socket.IO Not working anymore from Next.js 13.2.5-canary.27 to latest Next.js 13.4.1
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:41 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.16.0
      npm: 9.5.1
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.4.1-canary.2
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

warn  - Latest canary version not detected, detected: "13.4.1-canary.2", newest: "13.4.1".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/nextjs-socketio-u89y82?file=%2Fpackage.json&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A4%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A4%7D%5D

## Agenda of this project?
- To setup the Socket.Io server and share the information from one client to another user socket.io events

## The App is have the following things - 
-  An api for initialising the Socket.Io server - inside /pages/api/socketio
- A client page to connect to this server - inside /app/page

## Whats the Issue?
- There has been an unknown error in network connection from client side when its trying to create the Socket.Io connection.

## Since when this issue is happening?
- This has been happening from Next.js version 13.2.5-canary.27 version and its still happening even in latest canary release 13.4.1-canary.2 and also the latest release Next.js 13.4.1

## Do we have an Example project?
- Yes, I have created a testing project in codesandbox, that I have already shared. ([Link](https://codesandbox.io/p/sandbox/nextjs-socketio-u89y82?file=%2Fpackage.json&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A4%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A4%7D%5D))

## Steps to Reproduce
- Just see the console and network tabs for the client connection error. It will happen to all the versions on and above 13.2.5-canary.27
- And to see if its fixed, you need to downgrade to 13.2.5-canary.26

## Some References PR for / might this issue have created?
- I some how gathered the PR for the version when this Socket.Io broken (i.e. 13.2.5-canary.27) - [See here](https://github.com/vercel/next.js/pull/47208#issuecomment-1524565169)

### Describe the Bug

- There has been an unknown error in network connection from client side when its trying to create the Socket.Io connection.

### Expected Behavior

- There should be successful Socket.Io connection.
- It should exactly work the same way as it was working 13.2.5-canary.26 or lower.
- I was using 13.2.4 as stable version for our production application, as we had to use Socket.Io for websocket connection.

### Which browser are you using? (if relevant)

I am using latest Chrome version, but this has to do more with Next.js Core server. It has nothing to do with any browsers.

### How are you deploying your application? (if relevant)

Using Docker, but this is irrelevant too, as in localhost also, there is this issue.

<sub>[NEXT-1119](https://linear.app/vercel/issue/NEXT-1119/socketio-not-working-anymore-from-nextjs-1325-canary27-to-latest)</sub>

### PR DESCRIPTION
**Pull request by timneutkens**
Adds support for `getUpgradeHandler` with custom server. This API wasn't documented but it's crucial when you set up a custom socket server because otherwise there are two listeners for the same request.

Fixes #49334

### CODE REVIEW

**Comment by FFdhorkin** [comment]
I tried using your branch directly, but there appears to be changes to `useContext` in the canary branch that break my other code.

As such, I created a patch based on your PR and applied it to `v13.4.19`. It connects to websocket successfully now, so that's an improvement, but I get constant "client close with reason transport close" messages (and occasionally something about transport errors, but not as often).

<details>
  <summary>Click for detailed logs</summary>

```
upgrade from server
  socket.io:server incoming connection with id jlwUvqgFHbjYg8idAAAC +995ms
  socket.io-parser decoded 0 as {"type":0,"nsp":"/"} +990ms
  socket.io:client connecting to namespace / +974ms
  socket.io:namespace adding socket to nsp / +992ms
  socket.io:socket socket connected - writing packet +974ms
  socket.io:socket join room rU5yVeRre1s1V2dpAAAD +0ms
  socket.io-parser encoding packet {"type":0,"data":{"sid":"rU5yVeRre1s1V2dpAAAD"},"nsp":"/"} +2ms
  socket.io-parser encoded {"type":0,"data":{"sid":"rU5yVeRre1s1V2dpAAAD"},"nsp":"/"} as 0{"sid":"rU5yVeRre1s1V2dpAAAD"}
 +0ms
  socket.io:client client close with reason transport close +13ms
  socket.io:socket closing socket - reason transport close +12ms
upgrade from server
upgrade from server
upgrade from server
  socket.io:server incoming connection with id EaYvCRY5iEXegg1gAAAE +911ms
  socket.io-parser decoded 0 as {"type":0,"nsp":"/"} +913ms
  socket.io:client connecting to namespace / +902ms
  socket.io:namespace adding socket to nsp / +915ms
  socket.io:socket socket connected - writing packet +902ms
  socket.io:socket join room JkoFLXjb9ceZiOO7AAAF +0ms
  socket.io-parser encoding packet {"type":0,"data":{"sid":"JkoFLXjb9ceZiOO7AAAF"},"nsp":"/"} +0ms
  socket.io-parser encoded {"type":0,"data":{"sid":"JkoFLXjb9ceZiOO7AAAF"},"nsp":"/"} as 0{"sid":"JkoFLXjb9ceZiOO7AAAF"}
 +0ms
  socket.io:client client close with reason transport close +6ms
  socket.io:socket closing socket - reason transport close +6ms
upgrade from server
  socket.io:server incoming connection with id 9WbtQXHLbmiYcBZXAAAG +1s
  socket.io-parser decoded 0 as {"type":0,"nsp":"/"} +1s
  socket.io:client connecting to namespace / +1s
  socket.io:namespace adding socket to nsp / +1s
  socket.io:socket socket connected - writing packet +1s
  socket.io:socket join room 3I4cfaAoIBEyjbDWAAAH +0ms
  socket.io-parser encoding packet {"type":0,"data":{"sid":"3I4cfaAoIBEyjbDWAAAH"},"nsp":"/"} +0ms
  socket.io-parser encoded {"type":0,"data":{"sid":"3I4cfaAoIBEyjbDWAAAH"},"nsp":"/"} as 0{"sid":"3I4cfaAoIBEyjbDWAAAH"}
 +0ms
  socket.io:client client close with reason transport close +2ms
  socket.io:socket closing socket - reason transport close +2ms
```

</details>

I haven't been able to figure out how to get anything more useful out of the logs. The only real difference between my custom server and the one in the example in your PR is I have this as my server.js:

```js
// ...
const { Server: SocketIOServer } = require('socket.io');
let socketIOServer;
// ...

app.prepare().then(() => {
    const server = createServer(async (req, res) => {
        // ...
    });
    server.on('upgrade', (req, socket, head) => {
       console.log('upgrade from server');
       upgradeHandler(req, socket, head)
    })
    socketIOServer = new SocketIOServer(server, {
          path: '/api/socketio',
    };
});
```

I also tried adding `server.io = socketIOServer` but that didn't seem to work, either. Finally, I also tried using an endpoint in `pages/api/socketio.ts`, but that also didn't work.

I'm using Socket-IO `v4.7.2`. Everything works fine if I change `new SocketIOServer(server, {` to `new SocketIOServer(unusedPort, {` (and add CORS), so I suspect there's still something wrong after the fix in your PR

My client connect code (for test purposes) is 
```js
io(`ws://localhost:3000`, {
        path: '/api/socketio',
        transports: ['websocket'],
        upgrade: false,
    });
```

If I omit the transports/upgrade options, then it never exits the `polling` mode, even though I see the upgrade request go out. But it isn't constantly disconnecting/reconnecting. I see the HTTP 101 response in network logs, so I'm not sure what's going wrong:

<details>
  <summary>Click for detailed logs</summary>

```
  socket.io:server incoming connection with id 9z-OMDkDu-74LNz8AAAA +170ms
  socket.io-parser decoded 0 as {"type":0,"nsp":"/"} +0ms
  socket.io:client connecting to namespace / +0ms
  socket.io:namespace adding socket to nsp / +0ms
  socket.io:socket socket connected - writing packet +0ms
  socket.io:socket join room zLUpoPJyB4q1JUHwAAAB +0ms
  socket.io-parser encoding packet {"type":0,"data":{"sid":"zLUpoPJyB4q1JUHwAAAB"},"nsp":"/"} +1ms
  socket.io-parser encoded {"type":0,"data":{"sid":"zLUpoPJyB4q1JUHwAAAB"},"nsp":"/"} as 0{"sid":"zLUpoPJyB4q1JUHwAAAB"}
 +1ms
upgrade from server
- wait compiling /tictactoe/page (client and server)...
- event compiled client and server successfully in 1041 ms (688 modules)
  socket.io:client client close with reason transport error +2s
  socket.io:socket closing socket - reason transport error +2s
- wait compiling /favicon.ico/route (client and server)...
upgrade from server
- event compiled successfully in 482 ms (432 modules)
upgrade from server
  socket.io:server incoming connection with id So3KldW7lCzJS4T4AAAC +2s
  socket.io-parser decoded 0 as {"type":0,"nsp":"/"} +2s
  socket.io:client connecting to namespace / +845ms
  socket.io:namespace adding socket to nsp / +2s
  socket.io:socket socket connected - writing packet +846ms
  socket.io:socket join room Ce0u8CiIUNOcnRh2AAAD +1ms
  socket.io-parser encoding packet {"type":0,"data":{"sid":"Ce0u8CiIUNOcnRh2AAAD"},"nsp":"/"} +2ms
  socket.io-parser encoded {"type":0,"data":{"sid":"Ce0u8CiIUNOcnRh2AAAD"},"nsp":"/"} as 0{"sid":"Ce0u8CiIUNOcnRh2AAAD"}
 +0ms
```
</details>

File: packages/next/src/server/next.ts [hunk 1]
```diff
@@ -273,6 +273,13 @@ class NextCustomServer extends NextServer {
     }
   }
 
+  getUpgradeHandler() {
+    this.didWebSocketSetup = true
+    return async (req: IncomingMessage, socket: any, head: any) => {
+      this.upgradeHandler(req, socket, head)
+    }
+  }
```
Comment (styfle) [inline packages/next/src/server/next.ts:281]: Do we have tests for this behavior?

**Review by styfle — APPROVED** [review]

**Comment by gavro** [comment]
Any updates on this? Will this PR find it's way to v13.4.20? Looking forward to getting this fixed :)

**Comment by huozhi** [comment]
Close per https://github.com/vercel/next.js/issues/49334#issuecomment-1715511294

**Pull request closed without merging**